---
title: "Analyzing DNA Microarray Data in RStudio with Bioconductor and R"
author: "Sara Hussin"
date: "November 20, 2024"
output:
  pdf_document:
    number_sections: yes
    toc: yes
  html_document:
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
  word_document:
    toc: yes
editor_options:
  chunk_output_type: inline
urlcolor: blue
linkcolor: red
---

# Setup

Ensure that the following packages are installed: `knitr`.
```{r setup, echo=FALSE}
# This chunk installs and loads libraries required by the notebook
if (!requireNamespace("knitr", quietly = TRUE))
    install.packages('knitr')
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction to this Project

The purpose of this project is to gain experience in genome-wide analyses of differentially expressed genes using permutation distributions and the False Discovery Rate (FDR).

This project follows the first example provided in Chapter 16, "Two-Color Case Studies," of the Limma User Guide: Linear Models for Microarray and RNA-Seq Data User’s Guide, specifically section 16.1: "Swirl Zebrafish: A Single-Group Experiment." The example is augmented by generating some additional plots.

The Limma User Guide describes this data as follows:

> "**Background.** The experiment was carried out using zebrafish as a model organism to study the early development in vertebrates. Swirl is a point mutant in the BMP2 gene that affects the dorsal/ventral body axis. The main goal of the Swirl experiment is to identify genes with altered expression in the Swirl mutant compared to wild-type zebrafish.

> **The hybridizations.** Two sets of dye-swap experiments were performed making a total of four replicate hybridizations. Each of the arrays compares RNA from swirl fish with RNA from normal (“wild type”) fish."

# Download, Read In, and Inspect Swirl Data

We consulted the Limma User Guide (pages 76 and 77) to identify the "targets" file containing metadata about the samples hybridized to the four arrays in this experiment. The name of the file was added to the following code chunk, in the call to `readTargets`. This parametrized the subsequent call to `read.maimages`, which used filenames for image data from each microarray, returning a Limma `RGlist` object.

```{r}
setwd("swirl")
library(limma)
targets <- readTargets("SwirlSample.txt")
RG <- read.maimages(targets, source="spot")
RG
```

Similarly, on page 78, we identified the "GAL" file containing metadata about the genes spotted on the arrays. The file name was included in the following code chunk, in the call to `readGAL`.

```{r}
setwd("swirl")
RG$genes <- readGAL("fish.gal")
RG$printer <- getLayout(RG$genes)
```

# Create Image Plots

First, we created image plots for the background intensities of the first array, `swirl.1`.

```{r}
imageplot(log2(RG$Rb[,1]), RG$printer, low="white", high="red")
imageplot(log2(RG$Gb[,1]), RG$printer, low="white", high="green")
```

Next, we visualized how this background variation affects M-values computed for the first array. Note that no normalization method is applied when computing the M and A values here.

```{r}
MA_raw <- normalizeWithinArrays(RG, method="none")
imageplot(MA_raw$M[,1], RG$printer, zlim=c(-3,3))
```

Below is an interpretation of the result, as described on page 80 of the Limma User Guide:

> The `imageplot()` function lies the slide on its side, so the first print-tip group is bottom left in this plot. We can see a red streak across the middle two grids of the 3rd row caused by a scratch or dust on the array. Spots which are aﬀected by this artifact will have suspect M-values. The streak also shows up as darker regions in the background plots. The red streak seen on the image plot can be seen as a line of spots in the upper right of this plot.

# Visualize the Unnormalized Data Using an MA-plot and Boxplots of M-values by Print-Tip Group

We created an MA-plot for the first array and MA-plots grouped by print-tip for the unnormalized data:

```{r}
plotMD(MA_raw, column = 1) ## column = 1 is the default
abline(0,0,col="blue")
plotPrintTipLoess(MA_raw, array = 1)  ## array = 1 is the default
```

We also generated boxplots for all arrays to visualize the raw M-values without normalization:

```{r}
library(ggokabeito)
boxplot(MA_raw$M[,1]~MA_raw$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Raw M vs Print-Tip for swirl.1")
boxplot(MA_raw$M[,2]~MA_raw$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Raw M vs Print-Tip for swirl.2")
boxplot(MA_raw$M[,3]~MA_raw$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Raw M vs Print-Tip for swirl.3")
boxplot(MA_raw$M[,4]~MA_raw$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Raw M vs Print-Tip for swirl.4")
```

# Apply Default "print-tip" Normalization and Visualize M-values

We normalized within arrays using the default "printtiploess" method and visualized the normalized M-values.

```{r}
MA_print_tip <- normalizeWithinArrays(RG,method="printtiploess") 
# ^^^ method="printtiploess" is the default
plotMD(MA_print_tip)
abline(0,0,col="blue")
plotPrintTipLoess(MA_print_tip)
boxplot(MA_print_tip$M[,1]~MA_print_tip$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normlized M (printtiploess) vs Print-Tip for swirl.1")
boxplot(MA_print_tip$M[,2]~MA_print_tip$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normlized M (printtiploess) vs Print-Tip for swirl.2")
boxplot(MA_print_tip$M[,3]~MA_print_tip$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normlized M (printtiploess) vs Print-Tip for swirl.3")
boxplot(MA_print_tip$M[,4]~MA_print_tip$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normlized M (printtiploess) vs Print-Tip for swirl.4")
```

# Apply Scale Normalization Across Arrays

Following the Limma User Guide's recommendation, we applied scale normalization across arrays and visualized the results.

```{r}
MA_print_tip_scale <- normalizeBetweenArrays(MA_print_tip,method="scale")
plotMD(MA_print_tip_scale)
abline(0,0,col="blue")
plotPrintTipLoess(MA_print_tip_scale)
boxplot(MA_print_tip_scale$M[,1]~MA_print_tip_scale$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normlized M (printtiploess + scale) vs Print-Tip for swirl.1")
boxplot(MA_print_tip_scale$M[,2]~MA_print_tip_scale$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normlized M (printtiploess + scale) vs Print-Tip for swirl.2")
boxplot(MA_print_tip_scale$M[,3]~MA_print_tip_scale$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normalized M (printtiploess + scale)vs Print-Tip for swirl.3")
boxplot(MA_print_tip_scale$M[,4]~MA_print_tip_scale$genes$Block,xlab="Print-Tip",ylab="M",col=rep(palette_okabe_ito(1:8),each=2),main="Normalized M (printtiploess + scale) vs Print-Tip for swirl.4")
```

# Fit the Linear Model

```{r}
design <- modelMatrix(targets, ref="wild type")
design
fit_print_tip_scale <- lmFit(MA_print_tip_scale,design)
```

# Apply Empirical Bayes Moderation and Visualize Results

```{r}
ebayes_print_tip_scale <- eBayes(fit_print_tip_scale) ## the current function is eBayes
qqt(ebayes_print_tip_scale$t,
    df=ebayes_print_tip_scale$df.prior+ebayes_print_tip_scale$df.residual,pch=16,cex=0.2)
plotMD(ebayes_print_tip_scale)
abline(0,0,col="blue")
top30 <- order(ebayes_print_tip_scale$lods,decreasing=TRUE)[1:30]
text(ebayes_print_tip_scale$Amean[top30],ebayes_print_tip_scale$coef[top30],labels=ebayes_print_tip_scale$genes[top30,"Name"],cex=0.8,col="blue")
volcanoplot(ebayes_print_tip_scale,ylab="Surprisal (-logP)")
```

```{r}
ordinary.t <- fit_print_tip_scale$coef / fit_print_tip_scale$stdev.unscaled / fit_print_tip_scale$sigma
head(ordinary.t)
library(tidyverse)
tibble(moderated_t=unname(ebayes_print_tip_scale$t),ordinary_t=unname(ordinary.t)) |> mutate(shrink=(abs(moderated_t) < abs(ordinary_t))) |> ggplot(aes(x=ordinary_t,y=moderated_t,color=shrink)) + geom_point(alpha=0.5) + labs(title="Effect of Limma Moderation on Gene-Wise t") +  scale_color_okabe_ito()
```

# Identify Interesting Genes

```{r}
options(digits=3)
topTable(ebayes_print_tip_scale,number=30)
tt <- topTable(ebayes_print_tip_scale,n=Inf)
head(tt[tt$adj.P.Val <= 0.1,])
dim(tt[tt$adj.P.Val <= 0.1,])
dt <- decideTests(ebayes_print_tip_scale,p.value = 0.1)
summary(dt)
plotMD(ebayes_print_tip_scale,status=dt)
```

# Augment Results with FDR Estimation

We compared the number of differentially expressed genes (DEGs) identified by Limma to those obtained using FDR estimation methods, including pi0 estimation.

```{r}
library(FDRestimation)
hist(ebayes_print_tip_scale$p.value)

ebayes.fit.fdr <- p.fdr(pvalues=ebayes_print_tip_scale$p.value,threshold = 0.1)
plot(ebayes.fit.fdr, main="Benjamini-Hochberg FDRs for Swirl")
plot(ebayes.fit.fdr,xlim = c(0,1500),ylim=c(0,0.2),main="Blowup for Benjamini-Hochberg FDRs for Swirl")

sum(ebayes.fit.fdr$fdrs <= 0.1)
sum(ebayes.fit.fdr$`Results Matrix`$`BH FDRs`<= 0.1)
sum(ebayes.fit.fdr$`Results Matrix`$`Adjusted p-values`<= 0.1)
sum(ebayes.fit.fdr$`Reject Vector` == "Reject.H0")


ebayes.fit.fdr.set.pi0 <- p.fdr(pvalues=ebayes_print_tip_scale$p.value,
                                set.pi0 = get.pi0(ebayes_print_tip_scale$p.value),threshold = 0.1)
plot(ebayes.fit.fdr.set.pi0,main="Benjamini-Hochberg FDRs for Swirl with pi0 estimation")
plot(ebayes.fit.fdr.set.pi0,xlim = c(0,1500),ylim=c(0,0.2),main="Blowup for Benjamini-Hochberg FDRs for Swirl with pi0 estimation")


sum(ebayes.fit.fdr.set.pi0$fdrs <= 0.1)
sum(ebayes.fit.fdr.set.pi0$`Results Matrix`$`BH FDRs`<= 0.1)
sum(ebayes.fit.fdr.set.pi0$`Results Matrix`$`Adjusted p-values`<= 0.1)
sum(ebayes.fit.fdr.set.pi0$`Reject Vector` == "Reject.H0")
```